lib/rage/request.rb

```diff
@@ -14,6 +14,48 @@ def headers
     @headers ||= Headers.new(@env)
   end
 
+  # Check if the request is fresh.
+  # @param etag [String] The etag of the requested resource.
+  # @param last_modified [Time] The last modified time of the requested resource.
+  # @return [Boolean] True if the request is fresh, false otherwise.
+  # @example
+  #  request.fresh?(etag: "123", last_modified: Time.utc(2023, 12, 15))
+  #  request.fresh?(last_modified: Time.utc(2023, 12, 15))
+  #  request.fresh?(etag: "123")
+  def fresh?(etag:, last_modified:)
+    if_none_match = headers["HTTP_IF_NONE_MATCH"]
+    if_not_modified_since = headers["HTTP_IF_MODIFIED_SINCE"]
+
+    # Always render response when no freshness information
+    # is provided in the request.
+    return false unless if_none_match || if_not_modified_since
+
+    etag_matches?(
+      requested_etags: if_none_match, response_etag: etag
+    ) && not_modified?(
+      request_not_modified_since: if_not_modified_since,
+      response_last_modified: last_modified
+    )
+  end
+
+  private
+
+  def etag_matches?(requested_etags:, response_etag:)
+    requested_etags = requested_etags ? requested_etags.split(",").map(&:strip) : []
+
+    return true if requested_etags.empty?
+    return false if response_etag.nil?
+
+    requested_etags.include?(response_etag) || requested_etags.include?("*")
```

Comment:
Genuinely surprised this is much faster than calling `any?` 😯

Reply:
Just checked, the difference is indeed remarkable!
Hail to the ruby interpreter optimization magic 💯